pulser-simulation/pulser_simulation/hamiltonian.py

```diff
@@ -0,0 +1,640 @@
+from collections.abc import Mapping
+from dataclasses import asdict
+from typing import Any, Union, cast
+
+import numpy as np
+import qutip
+
+from pulser.devices._device_datacls import BaseDevice
+from pulser.register.base_register import QubitId
+from pulser.sampler.samples import SequenceSamples, _PulseTargetSlot
+from pulser_simulation.simconfig import SimConfig
+
+
+class Hamiltonian:
+    r"""Generates Hamiltonian from a sampled sequence and noise.
+
+    Args:
+        samples_obj: A sampled sequence whose ChannelSamples have same
+            duration.
+        qdict: A dictionary associating coordinates to qubit ids.
+        device: The device specifications.
+        sampling_rate: The fraction of samples that we wish to extract from
+            the samples to simulate. Has to be a value between 0.05 and 1.0.
+        config: Configuration to be used for this simulation.
+    """
+
+    def __init__(
+        self,
+        samples_obj: SequenceSamples,
+        qdict: dict[QubitId, np.ndarray],
+        device: BaseDevice,
+        sampling_rate: float,
+        config: SimConfig,
+    ) -> None:
+        """Instantiates a Hamiltonian object."""
+        self.samples_obj = samples_obj
+        self._qdict = qdict
+        self._device = device
+        self._sampling_rate = sampling_rate
+
+        # Type hints for attributes defined outside of __init__
+        self.basis_name: str
+        self._config: SimConfig
+        self.op_matrix: dict[str, qutip.Qobj]
+        self.basis: dict[str, qutip.Qobj]
+        self.dim: int
+        self._bad_atoms: dict[Union[str, int], bool] = {}
+        self._doppler_detune: dict[Union[str, int], float] = {}
+
+        # Define interaction
+        self._interaction = "XY" if self.samples_obj._in_xy else "ising"
+
+        # Initializing qubit infos
+        self._size = len(self._qdict)
+        self._qid_index = {qid: i for i, qid in enumerate(self._qdict)}
+
+        # Compute sampling times
+        self.duration = self.samples_obj.max_duration
+        self.sampling_times = self._adapt_to_sampling_rate(
+            # Include extra time step for final instruction from samples:
+            np.arange(self.duration, dtype=np.double)
+            / 1000
+        )
+
+        # Stores the qutip operators used in building the Hamiltonian
+        self.operators: dict[str, defaultdict[str, dict]] = {
+            addr: defaultdict(dict) for addr in ["Global", "Local"]
+        }
+        self._collapse_ops: list[qutip.Qobj] = []
+
+        self.set_config(config)
+
+    def _adapt_to_sampling_rate(self, full_array: np.ndarray) -> np.ndarray:
+        """Adapt list to correspond to sampling rate."""
+        indices = np.linspace(
+            0,
+            len(full_array) - 1,
+            int(self._sampling_rate * self.duration),
+            dtype=int,
+        )
+        return cast(np.ndarray, full_array[indices])
+
+    @property
+    def config(self) -> SimConfig:
```

Comment:
I'd actually keep all `config` related methods and properties in `QutipEmulator`